app/src/main/java/com/infomaniak/mail/ui/newMessage/NewMessageAiManager.kt

```diff
@@ -0,0 +1,246 @@
+    private val black by lazy { context.getColor(RCore.color.black) }
+
+    private var aiPromptFragment: AiPromptFragment? = null
+
+    private var valueAnimator: ValueAnimator? = null
+
+    fun initValues(
+        newMessageViewModel: NewMessageViewModel,
+        binding: FragmentNewMessageBinding,
+        fragment: NewMessageFragment,
+        aiViewModel: AiViewModel
+    ) {
+        super.initValues(newMessageViewModel, binding, fragment)
+        _aiViewModel = aiViewModel
+
+        viewLifecycleOwner.lifecycle.addObserver(LifecycleEventObserver { _: LifecycleOwner, event: Lifecycle.Event ->
+            if (event == Lifecycle.Event.ON_DESTROY) {
+                _aiViewModel = null
+                valueAnimator?.cancel()
+                valueAnimator = null
+            }
+        })
+    }
+
+    fun observeEverything() {
+        observeAiOutput()
+        observeAiPromptStatus()
+        observeAiFeatureFlagUpdates()
+    }
+
+    private fun observeAiOutput() = with(binding) {
+        aiViewModel.aiOutputToInsert.observe(viewLifecycleOwner) { (subject, content) ->
+            subject?.let { subjectTextField.setText(it) }
+            bodyText.setText(content)
+        }
+    }
+
+    private fun observeAiPromptStatus() {
+        aiViewModel.aiPromptOpeningStatus.observe(viewLifecycleOwner) { (shouldDisplay, shouldResetContent, becauseOfGeneration) ->
+            if (shouldDisplay) onAiPromptOpened(shouldResetContent) else onAiPromptClosed(becauseOfGeneration)
+        }
+    }
+
+    private fun onAiPromptOpened(resetPrompt: Boolean) = with(binding) {
+        if (resetPrompt) {
+            aiViewModel.apply {
+                aiPrompt = ""
+                aiPromptOpeningStatus.value?.shouldResetPrompt = false
+            }
+        }
+
+        // Keyboard is opened inside onCreate() of AiPromptFragment
+
+        val foundFragment = childFragmentManager.findFragmentByTag(AI_PROMPT_FRAGMENT_TAG) as? AiPromptFragment
+        if (aiPromptFragment == null) {
+            aiPromptFragment = foundFragment ?: AiPromptFragment()
+        }
+
+        if (foundFragment == null) {
+            childFragmentManager
+                .beginTransaction()
+                .add(aiPromptFragmentContainer.id, aiPromptFragment!!, AI_PROMPT_FRAGMENT_TAG)
+                .commitNow()
+        }
+
+        scrim.apply {
+            isVisible = true
+            isClickable = true
+        }
+
+        animateAiPrompt(true)
+        setAiPromptVisibility(true)
+        newMessageConstraintLayout.descendantFocusability = FOCUS_BLOCK_DESCENDANTS
+    }
+
+    private fun onAiPromptClosed(withoutTransition: Boolean) = with(binding) {
+
+        fun removeFragmentAndHideScrim() {
+            aiPromptFragment?.let {
+                childFragmentManager
+                    .beginTransaction()
+                    .remove(it)
+                    .commitNow()
+            }
+            aiPromptFragment = null
```

Comment:
Any reason to force the `aiPromptFragment` to be `null` ?

Reply:
Looking at the commits, it looks like @FabianDevel assigned it to `null` here to free its reference and avoid memory leaks. When navigating down the navigation tree I guess.

I do wonder if we could have a single instance of `AiPromptFragment()` though. But I think that's what I did go for when I implemented it back then and it actually caused an issue that said that the Fragment was already added when activity would get recreated. Maybe that part could use some comments?

Reply:
Yes it was because of leaks, as AiPropositionFragment also use this viewModel, but the AiPromptFragment is destroyed when we navigate to the proposition